### ISSUE TITLE
make watchdir retry interval configurable

### ISSUE THREAD

**Issue by freakynl** [transmission/transmission#1744]
Hi,

we're using Transmission on a Windows server with a watchdir. Users upload torrent files to a watchdir/pick-up dir.

Occasionally this doesn't work and the debug log says something along the lines of (see below).

Users are uploading these files from a laptop over a VPN. I think the virusscanners keep the files locked for just a bit too long in this situation. However I also think there's an issue with the code.

Preferably the retries would have some interval of a second or a couple of them in between. Now it seems like it detects the file, waits 1 sec (or well 998/999 millisec in many cases seeing the logs) and then logs 3 retries on the same exact millisecond before giving up.

Looking at https://github.com/transmission/transmission/blob/master/libtransmission/watchdir.c line 127 through 129 it would seem it was intended to have a 1 (to 10) sec interval.

I'm not a C/C++ dev however, I might be horribly wrong ;).

In any case, would it be possible to have a (larger/configurable) interval between the retries or some kind of larger delay between detection and actual pick-up? Would personally by far prefer an interval between retries of at least a second.

Files later added to the directory by someone else were picked up just fine and I moved these to local and back and they we're picked up fine then too.

--

[2021-07-21 08:59:47.899] watchdir Callback decided to retry file "XYZXYZ_NN08.torrent" (..\libtransmission\watchdir.c:110)
[2021-07-21 08:59:47.899] watchdir Callback decided to ignore file "XYZXYZ_NN08.torrent.mrq2vik.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 08:59:47.899] watchdir Callback decided to ignore file "XYZXYZ_NN08.torrent.mrq2vik.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 08:59:47.899] watchdir Callback decided to ignore file "XYZXYZ_NN08.torrent.mrq2vik.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 08:59:47.899] watchdir Callback decided to ignore file "XYZXYZ_NN08.torrent.mrq2vik.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:00:16.900] watchdir Callback decided to retry file "XYZXYZ_NN09.torrent" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:00:16.900] watchdir Callback decided to ignore file "XYZXYZ_NN09.torrent.u4urh6y.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:00:17.898] watchdir Callback decided to ignore file "XYZXYZ_NN09.torrent.u4urh6y.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:00:17.898] watchdir Callback decided to ignore file "XYZXYZ_NN09.torrent.u4urh6y.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:00:17.898] watchdir Callback decided to ignore file "XYZXYZ_NN09.torrent.u4urh6y.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:00:46.900] watchdir Callback decided to retry file "XYZXYZ_NN10.torrent" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:00:46.900] watchdir Callback decided to ignore file "XYZXYZ_NN10.torrent.k7xri11.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:00:47.898] watchdir Callback decided to ignore file "XYZXYZ_NN10.torrent.k7xri11.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:00:47.898] watchdir Callback decided to ignore file "XYZXYZ_NN10.torrent.k7xri11.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:00:47.898] watchdir Callback decided to ignore file "XYZXYZ_NN10.torrent.k7xri11.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:01:15.900] watchdir Callback decided to retry file "XYZXYZ_NN19.torrent" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:01:15.900] watchdir Callback decided to ignore file "XYZXYZ_NN19.torrent.px99e2u.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:01:16.899] watchdir Callback decided to ignore file "XYZXYZ_NN19.torrent.px99e2u.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:01:16.899] watchdir Callback decided to ignore file "XYZXYZ_NN19.torrent.px99e2u.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:01:16.899] watchdir Callback decided to ignore file "XYZXYZ_NN19.torrent.px99e2u.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:01:41.898] watchdir Callback decided to retry file "XYZXYZ_NN22.torrent" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:01:42.898] watchdir Callback decided to ignore file "XYZXYZ_NN22.torrent.gn2wz76.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:01:42.898] watchdir Callback decided to ignore file "XYZXYZ_NN22.torrent.gn2wz76.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:01:42.898] watchdir Callback decided to ignore file "XYZXYZ_NN22.torrent.gn2wz76.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:01:42.898] watchdir Callback decided to ignore file "XYZXYZ_NN22.torrent.gn2wz76.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:02:05.899] watchdir Callback decided to retry file "XYZXYZ_NN25.torrent" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:02:05.899] watchdir Callback decided to ignore file "XYZXYZ_NN25.torrent.jsyzi9k.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:02:05.899] watchdir Callback decided to ignore file "XYZXYZ_NN25.torrent.jsyzi9k.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:02:05.899] watchdir Callback decided to ignore file "XYZXYZ_NN25.torrent.jsyzi9k.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:02:05.899] watchdir Callback decided to ignore file "XYZXYZ_NN25.torrent.jsyzi9k.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:02:28.913] watchdir Callback decided to retry file "XYZXYZ_NN27.torrent" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:02:29.899] watchdir Callback decided to ignore file "XYZXYZ_NN27.torrent.4kh5d27.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:02:29.899] watchdir Callback decided to ignore file "XYZXYZ_NN27.torrent.4kh5d27.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:02:29.899] watchdir Callback decided to ignore file "XYZXYZ_NN27.torrent.4kh5d27.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:02:29.899] watchdir Callback decided to ignore file "XYZXYZ_NN27.torrent.4kh5d27.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:02:55.898] watchdir Callback decided to retry file "XYZXYZ_NN53.torrent" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:02:56.898] watchdir Callback decided to ignore file "XYZXYZ_NN53.torrent.d8bftbp.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:02:56.898] watchdir Callback decided to ignore file "XYZXYZ_NN53.torrent.d8bftbp.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:02:56.898] watchdir Callback decided to ignore file "XYZXYZ_NN53.torrent.d8bftbp.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:02:56.898] watchdir Callback decided to ignore file "XYZXYZ_NN53.torrent.d8bftbp.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:03:23.899] watchdir Callback decided to retry file "XYZXYZ_NN54.torrent" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:03:24.898] watchdir Callback decided to ignore file "XYZXYZ_NN54.torrent.rcjkdnd.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:03:24.898] watchdir Callback decided to ignore file "XYZXYZ_NN54.torrent.rcjkdnd.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:03:24.898] watchdir Callback decided to ignore file "XYZXYZ_NN54.torrent.rcjkdnd.partial" (..\libtransmission\watchdir.c:110)
[2021-07-21 09:03:24.898] watchdir Callback decided to ignore file "XYZXYZ_NN54.torrent.rcjkdnd.partial" (..\libtransmission\watchdir.c:110)

**Comment by freakynl** [comment]
Presume the user downloaded the file from browser directly to the watchdir. Browser adds the extension during download / safebrowsing/malware checks before renaming back.

At that point transmission won't try again.

Resolved it by using:
"watch-dir-force-generic": true
as found on https://forum.transmissionbt.com/viewtopic.php?t=17960

Think transmission is notified of new file creation, browser renames it and renamed it back. Renaming probably doesn't trigger a new file notification. At that point there are no more retries and it's not picked up thus (unless daemon is restarted).

watch-dir-force-generic checks the folder every 10 secs from what I got. Works like a charm now.

As this works fine I'm not sure if anyone still wants to look into it. I'll leave it open for you to decide.
